src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -475,6 +496,57 @@ entails SolverOptions{..} constraint context hints =
       pure (arg1, arg2, srcTypeLevelString (mkString $ h' <> t'))
     consSymbol _ _ _ = Nothing
 
+    solveIsReflectable :: [SourceType] -> Maybe [TypeClassDict]
+    solveIsReflectable [typeLevel, _] = do
+      (ref, typ) <- case typeLevel of
+        TypeLevelInt _ i -> pure (ReflectableInt i, tyInt)
+        TypeLevelString _ s -> pure (ReflectableString s, tyString)
+        _ -> Nothing
+      pure [TypeClassDictionaryInScope Nothing 0 (IsReflectableInstance ref) [] C.IsReflectable [] [] [typeLevel, typ] Nothing Nothing]
+    solveIsReflectable _ = Nothing
+
+    solveIntAdd :: [SourceType] -> Maybe [TypeClassDict]
+    solveIntAdd [arg0, arg1, arg2] = do
+      (arg0', arg1', arg2') <- addInts arg0 arg1 arg2
+      let args' = [arg0', arg1', arg2']
+      pure [TypeClassDictionaryInScope Nothing 0 EmptyClassInstance [] C.IntAdd [] [] args' Nothing Nothing]
+    solveIntAdd _ = Nothing
+
+    addInts :: SourceType -> SourceType -> SourceType -> Maybe (SourceType, SourceType, SourceType)
+    -- | l r -> o, l + r = o
+    addInts arg0@(TypeLevelInt _ l) arg1@(TypeLevelInt _ r) _ = pure (arg0, arg1, srcTypeLevelInt (l + r))
+    -- | l o -> r, o - l = r
+    addInts arg0@(TypeLevelInt _ l) _ arg2@(TypeLevelInt _ o) = pure (arg0, srcTypeLevelInt (o - l), arg2)
+    -- | r o -> l, o - r = l
+    addInts _ arg1@(TypeLevelInt _ r) arg2@(TypeLevelInt _ o) = pure (srcTypeLevelInt (o - r), arg1, arg2)
+    addInts _ _ _                                             = Nothing
+
+    solveIntCompare :: InstanceContext -> [SourceType] -> Maybe [TypeClassDict]
+    solveIntCompare _ [arg0@(TypeLevelInt _ a), arg1@(TypeLevelInt _ b), _] =
+      let ordering = case compare a b of
+            EQ -> C.orderingEQ
+            LT -> C.orderingLT
+            GT -> C.orderingGT
+          args' = [arg0, arg1, srcTypeConstructor ordering]
+      in pure [TypeClassDictionaryInScope Nothing 0 EmptyClassInstance [] C.IntCompare [] [] args' Nothing Nothing]
+    solveIntCompare ctx [a, b, c] = do
+      let compareDictsInScope = findDicts ctx C.IntCompare Nothing
+          givens = flip mapMaybe compareDictsInScope $ \case
+            dict | [a', b', c'] <- tcdInstanceTypes dict -> mkComparison a' b' c'
+                 | otherwise -> Nothing
+      solved <- solveComparison givens <$> mkComparison a b c
```

Comment:
Hm, we should be able to solve for unknown `c` in some contexts. Something like the following probably ought to work (modulo the fact that `IsReflectable Ordering` doesn't get solved):

```purs
compareIntTypes :: forall l r ord. Compare l r ord => IsReflectable Ordering ord => Proxy l -> Proxy r -> Ordering
compareIntTypes _ _ = reflectType (Proxy :: _ ord)

test :: forall m n p. Compare m n GT => Compare n p GT => Proxy m -> Proxy n -> Proxy p -> Ordering
test m _ p = compareIntTypes m p
```

Reply:
I could probably transform `solveRelation` into an inference-like function for this use-case then.

Reply:
I've done initial work on this locally but this case in particular seems like requires some the complexity removed from 6ddb8520a89c47af3885ddd140dc87924c3b8892 to do anything significant :confused:

Reply:
I'll take a look sometime soon; I'd be mildly surprised if that's the case but I could be glossing over something important.

Reply:
It turns out that it wasn't really the case in a43a2c71f96875059ac5295b7a32c729543f3b6e. A sort of compromise I had to make in the simpler framework was that `Unequal` actually means "less than" internally in the inequality graph, and "greater than" is solved by flipping the arguments around. So cases like:
```
a < b < c
x > y > z
```
Get normalized into:
```
a < b < c
z < y < x
```
If solving for `z -> x` solves with `LT` just fine but with `x -> z`, the search is flipped to do `z -> x` but it returns `GT` instead.